### DISCUSSION TITLE
The request signature we calculated does not match the signature you provided.

### DISCUSSION THREAD

**Discussion by booktrakker** [amzn/selling-partner-api-models#315] [Q&A]
We started getting this error after we rotated the LWA keys.  Developer support will not help us because we use the Swagger generated SDK in C#.

I have done one test in Postman, the query appears to work, however when we do it via code we get the following error (below)

Developer Support insists I need to fix the code, but I did not write it and it is beyond my ability.

What can we do?  We have 150 Sellers down for two weeks now with no resolution in sight.

{
  "errors": [
    {
      "message": "The request signature we calculated does not match the signature you provided. Check your AWS Secret Access Key and signing method. Consult the service documentation for details.

The Canonical String for this request should have been
'GET
/catalog/2022-04-01/items
identifiers=9780312978365&identifiersType=EAN&includedData=attributes%2Csummaries%2Cidentifiers%2CsalesRanks%2Cdimensions%2CproductTypes&marketplaceIds=ATVPDKIKX0DER&sellerId=A11TDGUOI15DHQ
accept:application/json
host:sellingpartnerapi-na.amazon.com
x-amz-access-token:Atc|MQEBIDiuLktXMLlAoxZK6jqYCv6GN4Tz7YNreMkAkYfJl0ZOfLudiNoxVLsTepN97vSFMKltaECGMqsIY9W54qh1yFi0zxOjEpwDAa3Lm9vCbIHJrme49USRuq_xrggs75RvNmFin_RGUg3rSFhaVhjg1ZeoBYKSzVx77Us5SpvENcYcp9uBTYITiA7ddvhDe_yK7uycsSR8npKA6CKWVEsfp8lnmwh4NhCTxIYHC699BbrfOeBCE6REexjyOCEmcVWTPbuaPEXZ5OTMOd5sk68ykMGnM3ixL_WtSGIa4IUoj89BC5RGkFdJ5zXhTkCUxT07L_PAUdyq_9fagPOlx1AkxO_botj6rpiZ1VYvgxNYF8-6RHM7PNNFjMF3IYrXyvKgMx4
x-amz-date:20230602T141827Z

accept;host;x-amz-access-token;x-amz-date
e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855'

The String-to-Sign should have been
'AWS4-HMAC-SHA256
20230602T141827Z     **<<<<<<-------  Date Amazon uses.**
20230602/us-east-1/execute-api/aws4_request
b6bfa0a4a43866b9fb74667d8a7a331b0a51fefb53dbd182fb2fbe7908f44287'
",
     "code": "InvalidSignature"
    }
  ]
}

From the SDK Code:

AWS4-HMAC-SHA256
20230602T142221Z     **<<<<<<<------- Date computed by Amazon SDK Code.**
20230602/us-east-1/execute-api/aws4_request
bed6fa8065cf26a66683ce2f513fc4af582b6bdb59a27b1859b30b3c8870e28b

**Comment by booktrakker** [comment]
I replaced the Signing files with new ones from the SDK site and everything started working.  Since this is an official Amazon SDK as I understand it, we need Developer Support to understand that those of us that use the SDK require support within the context of the SDK.

Our experience with Amazon Developer Support was horrible due to this disconnect.

**Comment by xbaha** [comment]
my app stopped too, but it was something wrong with id or secret.
make sure all the clientsIds start with "amzn1.application-oa2-client"
and all clientSecret start with "amzn1.oa2-cs.v1"

**Comment by booktrakker** [comment]
Thanks!

 

It turned out I needed to update the two files that process the signature…. Unfortunately Amazon Support is not up to speed with the SDKs and that complicates things considerably.

 

Jim Gutterman

Director of Development

BookTrakker.com

 ***@***.***> ***@***.***

**Comment by smargoli2** [comment]
So the C# Authorization SDK was updated but they didn't let anyone know...?

**Comment by booktrakker** [comment]
Seems that way…. It ties in with the complete lack of support from Developer Support.

 

I just went through two weeks of pure hell as a result.  I will be bringing this up at the Developer Summit.  It is totally unacceptable; we had 150 Sellers unable to use our product to manage their inventory.

 

Jim Gutterman

Director of Development

BookTrakker.com

 ***@***.***> ***@***.***

**Comment by SylvieManx** [comment]
hi, I saw your question and I would like to try to help you out with:
Are you able to use Postman requests? You need to sort the parameters in
alphabetical order before encrypting them in C#.

booktrakker ***@***.***> 于2023年6月5日周一 22:12写道：

**Comment by booktrakker** [comment]
Hello all,

It turned out that something did change in how the Signature is processed and I resolved it by updating the signature files from the SDK.  The truly aggravating part of all this is that Amazon Developer Support was completely useless and basically refused to support me effectively because they are wrapped around the API only axle and refuse to acknowledge the existence of the SDK in terms of providing support.

To say this is unacceptable would be a massive understatement.  I wasted two full weeks and then some where my Sellers were unable to use our product to manage their inventory and I only resolved it when I stumbled across this.

**Comment by xbaha** [comment]
many endpoints got updated without anyone getting notified, you will know when you get complaints or things stop working :)

**Comment by ericblade** [comment]
That sounds like something is fundamentally wrong with the procedures some are using.  Have had no issues like this with node SDK.  Renewed the keys, put them in the right place, and everything just kept working as usual.

**Comment by booktrakker** [comment]
I do not know what to say other than that until I replaced the two files I replaced it did not work.

 

The moment I replaced those two signing files it worked and has worked since.


I did not write the code in question so have nothing beyond that.

 

Jim Gutterman

Director of Development

BookTrakker.com

 ***@***.***> ***@***.***
